test/sim/misc/megaevolution.js

```diff
@@ -163,7 +163,7 @@ describe('Mega Evolution', function () {
 			assertLegalButCantMega('gen9nationaldexag@@@-ndag');
 
 			// don't add it where unnecessary
-			assert.false(Dex.formats.getRuleTable(Dex.formats.get('gen5anythinggoes')).has('megarayquazaclause'));
+			assert.false(Dex.formats.getRuleTable(Dex.formats.get('gen4anythinggoes')).has('megarayquazaclause'));
```

Comment:
Just like Gen 5 AG, Gen 4 AG is not guaranteed to exist in the future.

Can you fix this in a more reliable way?

Even just ``assert(Dex.formats.get('gen4anythinggoes').exists)`` can do the trick.

Reply:
Makes sense, made the change suggested in the later comment.

The code would now get this error message:
```
  2690 passing (5m)
  90 pending
  1 failing

  1) Mega Evolution
       Mega Rayquaza
         should implicitly add the Mega Rayquaza Clause when banned:
     Error: Unidentified format: gen5anythinggoes
      at TestTools.getFormat (test/common.js:47:30)
      at Context.<anonymous> (test/sim/misc/megaevolution.js:166:26)
      at processImmediate (node:internal/timers:476:21)

```

---

The other two instances of `formats.get` don't seem like they should / need to be changed.
```
$ grep -rn "formats.get(" ./test/
./test/random-battles/all-gens.js:31:	const format = Dex.formats.get(`${formatID}@@@Max Move Count = ${count}`);
./test/sim/team-validator/basic.js:11:	Dex.formats.getRuleTable(Dex.formats.get(format));
./test/common.js:46:	const format = Dex.formats.get(options.formatid);
./test/common.js:79:	format = Dex.formats.get(formatName);
```

Reply:
Something similar for helpers that wrap ``TeamValidator.get()``, probably? https://github.com/smogon/pokemon-showdown/blob/7078ff5142ef44b60289d2e41d566de4846712e8/test/random-battles/tools.js#L179-L180 

``if (!validator.format.exists) ...``

Reply:
How about checking `format.exists` in the `TeamValidator` constructor?
My understanding is that if you reached [this path](https://github.com/smogon/pokemon-showdown/blob/master/sim/dex-formats.ts#L687) in the `Format` constructor, then you get an empty ruletable, which probably isn't interesting for team validation.

I suppose someone could construct a populated Format and add an `exists: false` property, but I don't see why they would knowingly do that.

Reply:
Sounds good. Except that I have just remembered that ``format.exists`` is often a useless flag. ``format.effectType == 'Format'`` should be checked instead, to distinguish actual formats from rulesets. Other commits might need adjustments accordingly.

Reply:
I think both `exists` and `effectType` need to be checked, then. Invalid IDs yield `exists=false, effectType='Format'` and rulesets yield `exists=true, effectType={something else}`.

Annoyingly, even that wouldn't have caught all the bugs found in this PR.
* `./test/random-battles/all-gens.js` 166,15-29 | uses `Dex.forFormat()`, bypassing the checks in `common.getFormat`
* the '*' tag thing was a silent bug because the related code does a falsy check against the field...
```
if ('!+-'.includes(rule.charAt(0))) continue;    // didn't handle * case
const subformat = dex.formats.get(rule);    // get exists=false format
if (subformat.onValidateTeam && ruleTable.has(subformat.id)) {    // field DNE, skip
        problems = problems.concat(subformat.onValidateTeam.call(this, team, format, teamHas) || []);
}
```

---

I wouldn't be shocked if the `exists: false` idiom in getters is currently hiding dozens of bugs.
Default values are nice, but such cases are probably worth being explicit about. Oh well.

Reply:
While that may be true, these randbats test are also terribly sloppy. Test writers should realize when they add tests for obviously temporary formats such as CAP's or baby randbats.

https://github.com/smogon/pokemon-showdown/blob/e9670c84cc71e383c4ff70659c02a47762106291/test/random-battles/all-gens.js#L159-L162

In that case, it's wild to reach to call ``Dex.forFormat(format)`` when ``Dex.mod(mod)`` is as readily available. But, still, the preferred version for tests should be ``common.mod(mod)``.

Reply:
Ok, to be fair, you really can't blame them for the parsing. That's probably a symptom of the enemy stand, 「[DexStats Requiem](https://github.com/smogon/pokemon-showdown/pull/10619#issuecomment-2412539312)」.
`ModdedDex::gen` isn't updated until you call `loadData`. 
It's a footgun.

*grumbling about predictable data flow*

Reply:
To respond to this part, 
> terribly sloppy. Test writers should realize

While that could be said for any individual instance, in general I don't think that 'be careful' is a scalable approach to correctness, security or performance in a large system. There's just too much code for anybody to really understand all of it.

<details><summary>Pop Quiz</summary>
What does this print?

```ts
console.log(Dex.species.get('something invalid').effectType);
console.log(Dex.items.get('something invalid').effectType);
console.log(Dex.moves.get('something invalid').effectType);
console.log(Dex.abilities.get('something invalid').effectType);
console.log(Dex.formats.get('something invalid').effectType);
console.log(Dex.types.get('something invalid').effectType);
console.log(Dex.natures.get('something invalid').effectType);
```

<details><summary>answer</summary>

```
Pokemon
Item
Move
Ability
Condition    <----- Did anybody expect this? Should they have?
EffectType
Nature
```
</details>
</details>

I think the assertions being added are a more effective strategy.

Reply:
``Condition`` is PS's equivalent to JavaScript's equivalent to ``Object``. In JS, everything is an Object. In PS, everything is a ``Condition`` (formerly called ``Effect``). Given the ambiguity between formats, rulesets, and other stuff, ``Condition`` is indeed fitting. The ``Condition`` semantics are further elaborated by the fact that ``Format`` and ``Ruleset`` may and often include event handlers.

Of course, this is likely surprising for many, but it's a direct consequence of the design or lack thereof of PS format system (not favoring either side, I can see the benefits of the current system, though certainly it can be improved).

Also, these lines at ``dex-formats.ts`` don't help, which are certainly incorrect: 
```ts
type FormatEffectType = 'Format' | 'Ruleset' | 'Rule' | 'ValidatorRule';
this.effectType = Utils.getString(data.effectType) as FormatEffectType || 'Format';
```

Having said all that, I did forget these facts at the time of reading your quiz, because I have actually been a very inactive contributor. But, hey, that was the reason why I remembered and suggested that we should just check that ``effectType`` is ``'Format'``. It's indeed enough ;D

Reply:
Right. Since you are interested in the findEventHandlers situation, that link should be very useful to you.

Reply:
So I looked and it seems that the effectType convention for invalid Formats dates back to the [start of the repo](https://github.com/smogon/pokemon-showdown/blob/4ce2d9503769e57aa6abec5295d4f477e54b3fe5/tools.js#L110-L130) and the introduction of [getFormat](https://github.com/smogon/pokemon-showdown/blob/b3054fefdf59674ac8295d73ff8381f6cadadaaa/tools.js#L249). Ah well. I just feel bad for the next person who has to infer why a `Format` sometimes isn't a 'Format'.

---

I ended up going with `common.getFormat` because that would have caught the expired CAP case, and `gen9baby` isn't a mod so that would crash.

<details><summary>Original</summary>

```ts
for (const format of Object.keys(formatInfo)) {
        const filename = formatInfo[format].filename;
        const setsJSON = require(`../../dist/data/random-battles/${filename}.json`);
        const mod = filename.split('/')[0];
        const genNum = parseInt(mod[3]);
        const rounds = 100;
        const dex = Dex.forFormat(format);
        console.log(format, genNum, dex.gen);
```
=>
```
gen2randombattle 2 2
gen3randombattle 3 3
gen4randombattle 4 4
gen5randombattle 5 5
gen6randombattle 6 6
gen7randombattle 7 7
gen9randombattle 9 9
gen9randomdoublesbattle 9 9
gen9babyrandombattle 9 9
```
</details>

---

<details><summary>v2 (to demonstrate footgun in getFormat)</summary>

```ts
for (const format of Object.keys(formatInfo)) {
        const filename = formatInfo[format].filename;
        const setsJSON = require(`../../dist/data/random-battles/${filename}.json`);
        const dex = common.mod(common.getFormat(format).mod).dex; // .getFormat verifies format exists
        console.log(format, dex.gen);
        const rounds = 100;
```
=>
```ts
gen2randombattle 9
gen3randombattle 9
gen4randombattle 9
gen5randombattle 9
gen6randombattle 9
gen7randombattle 9
gen9randombattle 9
gen9randomdoublesbattle 9
gen9babyrandombattle 9
```

Fails in a very non-obvious fashion.
</details>

---

<details><summary>v3</summary>

```ts
for (const format of Object.keys(formatInfo)) {
        const filename = formatInfo[format].filename;
        const setsJSON = require(`../../dist/data/random-battles/${filename}.json`);
        const dex = common.mod(common.getFormat({formatid: format}).mod).dex; // verifies format exists
        const genNum = dex.gen;
        console.log(format, genNum);
        const rounds = 100;
```
=>
```
gen2randombattle 2
gen3randombattle 3
gen4randombattle 4
gen5randombattle 5
gen6randombattle 6
gen7randombattle 7
gen9randombattle 9
gen9randomdoublesbattle 9
gen9babyrandombattle 9
```
</details>

---

Reply:
Feel free to use v3 if you prefer it. But it looks like ``common.mod`` should also be calling ``loadData()`` itself. I am not sure what the state of your other PRs is, so maybe that can be done later, to avoid conflicts.

Reply:
While we are still doing archaeology, I tracked the ``gen`` loading issue to 3716f360ffc15.

> This basically means you don't need to do Tools.includeData() or anything like that anymore

Is TypeScript full of void promises??!